light-node/src/commands/start.rs

```diff
@@ -4,94 +4,106 @@
 /// accessors along with logging macros. Customize as you see fit.
 use abscissa_core::{config, Command, FrameworkError, Options, Runnable};
 use std::process;
-use std::time::{Duration, SystemTime};
 
-use tendermint::hash;
-use tendermint::lite;
-use tendermint::lite::error::Error;
-use tendermint::lite::ValidatorSet as _;
-use tendermint::lite::{Header, Height, Requester, TrustThresholdFraction};
-use tendermint::Hash;
-
-use tendermint_rpc as rpc;
-
-use crate::application::APPLICATION;
-use crate::config::LightNodeConfig;
-use crate::prelude::*;
-use crate::requester::RPCRequester;
-use crate::store::{MemStore, State};
+use crate::application::{app_config, APPLICATION};
+use crate::config::{LightClientConfig, LightNodeConfig};
+use crate::rpc::Server;
+
+use crate::rpc;
+use abscissa_core::path::PathBuf;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::ops::Deref;
+use std::time::Duration;
+use tendermint_light_client::components::clock::SystemClock;
+use tendermint_light_client::components::io::ProdIo;
+use tendermint_light_client::components::scheduler;
+use tendermint_light_client::components::verifier::ProdVerifier;
+use tendermint_light_client::evidence::ProdEvidenceReporter;
+use tendermint_light_client::fork_detector::ProdForkDetector;
+use tendermint_light_client::light_client;
+use tendermint_light_client::light_client::LightClient;
+use tendermint_light_client::peer_list::{PeerList, PeerListBuilder};
+use tendermint_light_client::state::State;
+use tendermint_light_client::store::sled::SledStore;
+use tendermint_light_client::store::LightStore;
+use tendermint_light_client::supervisor::Handle;
+use tendermint_light_client::supervisor::{Instance, Supervisor};
+use tendermint_light_client::types::Status;
 
 /// `start` subcommand
 ///
-/// The `Options` proc macro generates an option parser based on the struct
-/// definition, and is defined in the `gumdrop` crate. See their documentation
-/// for a more comprehensive example:
-///
-/// <https://docs.rs/gumdrop/>
 #[derive(Command, Debug, Options)]
 pub struct StartCmd {
-    /// RPC address to request headers and validators from.
-    #[options(free)]
-    rpc_addr: String,
+    /// Path to configuration file
+    #[options(
+        short = "l",
+        long = "listen",
+        help = "address the rpc server will serve"
+    )]
+    pub listen_addr: Option<SocketAddr>,
+
+    /// Path to configuration file
+    #[options(short = "c", long = "config", help = "path to light_node.toml")]
+    pub config: Option<PathBuf>,
 }
 
 impl Runnable for StartCmd {
     /// Start the application.
     fn run(&self) {
         if let Err(err) = abscissa_tokio::run(&APPLICATION, async {
-            let config = app_config();
+            StartCmd::assert_init_was_run();
```

Comment:
Generally I think the run method is a bit bloated, it would be nice to break out the nitty-gritty of the setup and just see a flow like:

* constructs supervisor
* constructs rpc
* spawns supervisor run loop
* spawns rpc
* runs verify continously

Reply:
Absolutely. I was just eager to test against a running tendermint node as quickly as possible. Will refactor this before setting it to ready for review.

Reply:
Dope!

Reply:
Does that look better now? https://github.com/informalsystems/tendermint-rs/pull/430/commits/08ce5cc1f6fe63b8f176a8a4447782cc27ac5574